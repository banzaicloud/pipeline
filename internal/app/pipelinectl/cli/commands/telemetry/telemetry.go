// Copyright Â© 2020 Banzai Cloud
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package telemetry

import (
	"crypto/tls"
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"os"
	"time"

	"emperror.dev/errors"
	"github.com/MakeNowJust/heredoc"
	dto "github.com/prometheus/client_model/go"
	"github.com/prometheus/prom2json"
	"github.com/spf13/cobra"
	"github.com/spf13/viper"
)

type options struct {
	telemetryUrl string
	verify       bool

	// for tests
	telemetryFile string
}

func NewTelemetryCommand() *cobra.Command {
	options := options{}
	cmd := &cobra.Command{
		Use:   "telemetry",
		Short: "Get telemetry information",
		Long: heredoc.Doc(`
			Get pipeline telemetry information
		`),
		Args: cobra.MaximumNArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			options.telemetryUrl = viper.GetString("telemetry.url")
			options.verify = viper.GetBool("telemetry.verify")

			cmd.SilenceErrors = true
			cmd.SilenceUsage = true

			status, err := getTelemetry(options)
			if err != nil {
				return err
			}

			metrics, err := json.Marshal(status)
			if err != nil {
				return errors.WrapIff(err, "unable to marshal metrics %+v", metrics)
			}

			fmt.Println(string(metrics))

			return nil
		},
	}

	return cmd
}

func getTelemetry(options options) ([]*prom2json.Family, error) {
	mfChan := make(chan *dto.MetricFamily, 1024)

	if options.telemetryFile != "" {
		input, err := os.Open(options.telemetryFile)
		if err != nil {
			return nil, errors.WrapIf(err, "opening telemetry file")
		}
		go func() {
			if err := prom2json.ParseReader(input, mfChan); err != nil {
				log.Fatal("error reading metrics:", err)
			}
		}()
	} else {
		transport, err := makeTransport(options.verify)
		if err != nil {
			return nil, errors.WrapIf(err, "creating transport")
		}
		go func() {
			err := prom2json.FetchMetricFamilies(options.telemetryUrl, mfChan, transport)
			if err != nil {
				log.Fatalln(err)
			}
		}()
	}

	result := []*prom2json.Family{}
	for mf := range mfChan {
		result = append(result, prom2json.NewFamily(mf))
	}

	return result, nil
}

func makeTransport(skipServerCertCheck bool) (*http.Transport, error) {
	transport := http.DefaultTransport.(*http.Transport).Clone()
	transport.ResponseHeaderTimeout = time.Second
	tlsConfig := &tls.Config{InsecureSkipVerify: skipServerCertCheck}
	transport.TLSClientConfig = tlsConfig
	return transport, nil
}
