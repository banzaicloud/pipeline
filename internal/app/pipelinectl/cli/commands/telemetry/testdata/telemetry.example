# HELP go_gc_duration_seconds A summary of the pause duration of garbage collection cycles.
# TYPE go_gc_duration_seconds summary
go_gc_duration_seconds{quantile="0"} 2.3533e-05
go_gc_duration_seconds{quantile="0.25"} 3.5095e-05
go_gc_duration_seconds{quantile="0.5"} 6.0324e-05
go_gc_duration_seconds{quantile="0.75"} 0.000229192
go_gc_duration_seconds{quantile="1"} 0.000241134
go_gc_duration_seconds_sum 0.00081884
go_gc_duration_seconds_count 7
# HELP go_goroutines Number of goroutines that currently exist.
# TYPE go_goroutines gauge
go_goroutines 45
# HELP go_info Information about the Go environment.
# TYPE go_info gauge
go_info{version="go1.14.3"} 1
# HELP go_memstats_alloc_bytes Number of bytes allocated and still in use.
# TYPE go_memstats_alloc_bytes gauge
go_memstats_alloc_bytes 1.0304632e+07
# HELP go_memstats_alloc_bytes_total Total number of bytes allocated, even if freed.
# TYPE go_memstats_alloc_bytes_total counter
go_memstats_alloc_bytes_total 2.4473024e+07
# HELP go_memstats_buck_hash_sys_bytes Number of bytes used by the profiling bucket hash table.
# TYPE go_memstats_buck_hash_sys_bytes gauge
go_memstats_buck_hash_sys_bytes 1.456805e+06
# HELP go_memstats_frees_total Total number of frees.
# TYPE go_memstats_frees_total counter
go_memstats_frees_total 163531
# HELP go_memstats_gc_cpu_fraction The fraction of this program's available CPU time used by the GC since the program started.
# TYPE go_memstats_gc_cpu_fraction gauge
go_memstats_gc_cpu_fraction 0.0008389383820980088
# HELP go_memstats_gc_sys_bytes Number of bytes used for garbage collection system metadata.
# TYPE go_memstats_gc_sys_bytes gauge
go_memstats_gc_sys_bytes 3.590408e+06
# HELP go_memstats_heap_alloc_bytes Number of heap bytes allocated and still in use.
# TYPE go_memstats_heap_alloc_bytes gauge
go_memstats_heap_alloc_bytes 1.0304632e+07
# HELP go_memstats_heap_idle_bytes Number of heap bytes waiting to be used.
# TYPE go_memstats_heap_idle_bytes gauge
go_memstats_heap_idle_bytes 5.267456e+07
# HELP go_memstats_heap_inuse_bytes Number of heap bytes that are in use.
# TYPE go_memstats_heap_inuse_bytes gauge
go_memstats_heap_inuse_bytes 1.35168e+07
# HELP go_memstats_heap_objects Number of allocated objects.
# TYPE go_memstats_heap_objects gauge
go_memstats_heap_objects 59303
# HELP go_memstats_heap_released_bytes Number of heap bytes released to OS.
# TYPE go_memstats_heap_released_bytes gauge
go_memstats_heap_released_bytes 5.0774016e+07
# HELP go_memstats_heap_sys_bytes Number of heap bytes obtained from system.
# TYPE go_memstats_heap_sys_bytes gauge
go_memstats_heap_sys_bytes 6.619136e+07
# HELP go_memstats_last_gc_time_seconds Number of seconds since 1970 of last garbage collection.
# TYPE go_memstats_last_gc_time_seconds gauge
go_memstats_last_gc_time_seconds 1.5941881536458519e+09
# HELP go_memstats_lookups_total Total number of pointer lookups.
# TYPE go_memstats_lookups_total counter
go_memstats_lookups_total 0
# HELP go_memstats_mallocs_total Total number of mallocs.
# TYPE go_memstats_mallocs_total counter
go_memstats_mallocs_total 222834
# HELP go_memstats_mcache_inuse_bytes Number of bytes in use by mcache structures.
# TYPE go_memstats_mcache_inuse_bytes gauge
go_memstats_mcache_inuse_bytes 13888
# HELP go_memstats_mcache_sys_bytes Number of bytes used for mcache structures obtained from system.
# TYPE go_memstats_mcache_sys_bytes gauge
go_memstats_mcache_sys_bytes 16384
# HELP go_memstats_mspan_inuse_bytes Number of bytes in use by mspan structures.
# TYPE go_memstats_mspan_inuse_bytes gauge
go_memstats_mspan_inuse_bytes 257312
# HELP go_memstats_mspan_sys_bytes Number of bytes used for mspan structures obtained from system.
# TYPE go_memstats_mspan_sys_bytes gauge
go_memstats_mspan_sys_bytes 278528
# HELP go_memstats_next_gc_bytes Number of heap bytes when next garbage collection will take place.
# TYPE go_memstats_next_gc_bytes gauge
go_memstats_next_gc_bytes 1.6738688e+07
# HELP go_memstats_other_sys_bytes Number of bytes used for other system allocations.
# TYPE go_memstats_other_sys_bytes gauge
go_memstats_other_sys_bytes 1.817939e+06
# HELP go_memstats_stack_inuse_bytes Number of bytes in use by the stack allocator.
# TYPE go_memstats_stack_inuse_bytes gauge
go_memstats_stack_inuse_bytes 917504
# HELP go_memstats_stack_sys_bytes Number of bytes obtained from system for stack allocator.
# TYPE go_memstats_stack_sys_bytes gauge
go_memstats_stack_sys_bytes 917504
# HELP go_memstats_sys_bytes Number of bytes obtained from system.
# TYPE go_memstats_sys_bytes gauge
go_memstats_sys_bytes 7.4268928e+07
# HELP go_threads Number of OS threads created.
# TYPE go_threads gauge
go_threads 12
# HELP pipeline_cluster_total the number of clusters launched
# TYPE pipeline_cluster_total counter
pipeline_cluster_total{location="eu-west-1",provider="amazon"} 10
# HELP pipeline_drain_mode read only mode is on/off
# TYPE pipeline_drain_mode gauge
pipeline_drain_mode 0
# HELP pipeline_request_duration_seconds The HTTP request latencies in seconds.
# TYPE pipeline_request_duration_seconds summary
pipeline_request_duration_seconds_sum 0
pipeline_request_duration_seconds_count 0
# HELP pipeline_request_size_bytes The HTTP request sizes in bytes.
# TYPE pipeline_request_size_bytes summary
pipeline_request_size_bytes_sum 0
pipeline_request_size_bytes_count 0
# HELP pipeline_response_size_bytes The HTTP response sizes in bytes.
# TYPE pipeline_response_size_bytes summary
pipeline_response_size_bytes_sum 0
pipeline_response_size_bytes_count 0
# HELP promhttp_metric_handler_requests_in_flight Current number of scrapes being served.
# TYPE promhttp_metric_handler_requests_in_flight gauge
promhttp_metric_handler_requests_in_flight 1
# HELP promhttp_metric_handler_requests_total Total number of scrapes by HTTP status code.
# TYPE promhttp_metric_handler_requests_total counter
promhttp_metric_handler_requests_total{code="200"} 0
promhttp_metric_handler_requests_total{code="500"} 0
promhttp_metric_handler_requests_total{code="503"} 0
