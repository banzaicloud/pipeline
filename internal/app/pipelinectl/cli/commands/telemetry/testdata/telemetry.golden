[{"name":"go_gc_duration_seconds","help":"A summary of the pause duration of garbage collection cycles.","type":"SUMMARY","metrics":[{"quantiles":{"0":"2.3533e-05","0.25":"3.5095e-05","0.5":"6.0324e-05","0.75":"0.000229192","1":"0.000241134"},"count":"7","sum":"0.00081884"}]},{"name":"go_goroutines","help":"Number of goroutines that currently exist.","type":"GAUGE","metrics":[{"value":"45"}]},{"name":"go_info","help":"Information about the Go environment.","type":"GAUGE","metrics":[{"labels":{"version":"go1.14.3"},"value":"1"}]},{"name":"go_memstats_alloc_bytes","help":"Number of bytes allocated and still in use.","type":"GAUGE","metrics":[{"value":"1.0304632e+07"}]},{"name":"go_memstats_alloc_bytes_total","help":"Total number of bytes allocated, even if freed.","type":"COUNTER","metrics":[{"value":"2.4473024e+07"}]},{"name":"go_memstats_buck_hash_sys_bytes","help":"Number of bytes used by the profiling bucket hash table.","type":"GAUGE","metrics":[{"value":"1.456805e+06"}]},{"name":"go_memstats_frees_total","help":"Total number of frees.","type":"COUNTER","metrics":[{"value":"163531"}]},{"name":"go_memstats_gc_cpu_fraction","help":"The fraction of this program's available CPU time used by the GC since the program started.","type":"GAUGE","metrics":[{"value":"0.0008389383820980088"}]},{"name":"go_memstats_gc_sys_bytes","help":"Number of bytes used for garbage collection system metadata.","type":"GAUGE","metrics":[{"value":"3.590408e+06"}]},{"name":"go_memstats_heap_alloc_bytes","help":"Number of heap bytes allocated and still in use.","type":"GAUGE","metrics":[{"value":"1.0304632e+07"}]},{"name":"go_memstats_heap_idle_bytes","help":"Number of heap bytes waiting to be used.","type":"GAUGE","metrics":[{"value":"5.267456e+07"}]},{"name":"go_memstats_heap_inuse_bytes","help":"Number of heap bytes that are in use.","type":"GAUGE","metrics":[{"value":"1.35168e+07"}]},{"name":"go_memstats_heap_objects","help":"Number of allocated objects.","type":"GAUGE","metrics":[{"value":"59303"}]},{"name":"go_memstats_heap_released_bytes","help":"Number of heap bytes released to OS.","type":"GAUGE","metrics":[{"value":"5.0774016e+07"}]},{"name":"go_memstats_heap_sys_bytes","help":"Number of heap bytes obtained from system.","type":"GAUGE","metrics":[{"value":"6.619136e+07"}]},{"name":"go_memstats_last_gc_time_seconds","help":"Number of seconds since 1970 of last garbage collection.","type":"GAUGE","metrics":[{"value":"1.5941881536458519e+09"}]},{"name":"go_memstats_lookups_total","help":"Total number of pointer lookups.","type":"COUNTER","metrics":[{"value":"0"}]},{"name":"go_memstats_mallocs_total","help":"Total number of mallocs.","type":"COUNTER","metrics":[{"value":"222834"}]},{"name":"go_memstats_mcache_inuse_bytes","help":"Number of bytes in use by mcache structures.","type":"GAUGE","metrics":[{"value":"13888"}]},{"name":"go_memstats_mcache_sys_bytes","help":"Number of bytes used for mcache structures obtained from system.","type":"GAUGE","metrics":[{"value":"16384"}]},{"name":"go_memstats_mspan_inuse_bytes","help":"Number of bytes in use by mspan structures.","type":"GAUGE","metrics":[{"value":"257312"}]},{"name":"go_memstats_mspan_sys_bytes","help":"Number of bytes used for mspan structures obtained from system.","type":"GAUGE","metrics":[{"value":"278528"}]},{"name":"go_memstats_next_gc_bytes","help":"Number of heap bytes when next garbage collection will take place.","type":"GAUGE","metrics":[{"value":"1.6738688e+07"}]},{"name":"go_memstats_other_sys_bytes","help":"Number of bytes used for other system allocations.","type":"GAUGE","metrics":[{"value":"1.817939e+06"}]},{"name":"go_memstats_stack_inuse_bytes","help":"Number of bytes in use by the stack allocator.","type":"GAUGE","metrics":[{"value":"917504"}]},{"name":"go_memstats_stack_sys_bytes","help":"Number of bytes obtained from system for stack allocator.","type":"GAUGE","metrics":[{"value":"917504"}]},{"name":"go_memstats_sys_bytes","help":"Number of bytes obtained from system.","type":"GAUGE","metrics":[{"value":"7.4268928e+07"}]},{"name":"go_threads","help":"Number of OS threads created.","type":"GAUGE","metrics":[{"value":"12"}]},{"name":"pipeline_cluster_total","help":"the number of clusters launched","type":"COUNTER","metrics":[{"labels":{"location":"eu-west-1","provider":"amazon"},"value":"10"}]},{"name":"pipeline_drain_mode","help":"read only mode is on/off","type":"GAUGE","metrics":[{"value":"0"}]},{"name":"pipeline_request_duration_seconds","help":"The HTTP request latencies in seconds.","type":"SUMMARY","metrics":[{"count":"0","sum":"0"}]},{"name":"pipeline_request_size_bytes","help":"The HTTP request sizes in bytes.","type":"SUMMARY","metrics":[{"count":"0","sum":"0"}]},{"name":"pipeline_response_size_bytes","help":"The HTTP response sizes in bytes.","type":"SUMMARY","metrics":[{"count":"0","sum":"0"}]},{"name":"promhttp_metric_handler_requests_in_flight","help":"Current number of scrapes being served.","type":"GAUGE","metrics":[{"value":"1"}]},{"name":"promhttp_metric_handler_requests_total","help":"Total number of scrapes by HTTP status code.","type":"COUNTER","metrics":[{"labels":{"code":"200"},"value":"0"},{"labels":{"code":"500"},"value":"0"},{"labels":{"code":"503"},"value":"0"}]}]